pkg/testing/runner/debian.go

```diff
@@ -140,72 +140,72 @@ func (DebianRunner) Run(ctx context.Context, verbose bool, c *ssh.Client, logger
-		session.Stderr = newPrefixOutput(os.Stderr, fmt.Sprintf(">>> (%s) Test output (stderr): ", logger.Prefix()))
-		session.Stdin = execTest
-		// allowed to fail because tests might fail
-		_ = session.Run("bash")
-		_ = session.Close()
-
-		// fetch the contents for each package
-		for _, pkg := range batch.Tests {
-			resultPkg, err := getRunnerPackageResult(ctx, c, pkg, prefix)
-			if err != nil {
-				return OSRunnerResult{}, err
-			}
-			result.Packages = append(result.Packages, resultPkg)
+			return OSRunnerResult{}, fmt.Errorf("error running non-sudo tests: %w", err)
 		}
+		result.Packages = results
 	}
 
 	if len(sudoTests) > 0 {
 		prefix := fmt.Sprintf("%s-sudo", prefix)
 		vars := fmt.Sprintf(`GOPATH="$HOME/go" PATH="$HOME/go/bin:$PATH" AGENT_VERSION="%s" TEST_DEFINE_PREFIX="%s" TEST_DEFINE_TESTS="%s"`, agentVersion, prefix, strings.Join(sudoTests, ","))
 		vars = extendVars(vars, env)
-		logger.Logf("Starting sudo tests")
-		logArg := ""
-		if verbose {
-			logArg = "-v"
-		}
 		script := fmt.Sprintf(`cd agent && sudo %s ~/go/bin/mage %s integration:testOnRemote`, vars, logArg)
-		execTest := strings.NewReader(script)
 
-		session, err := c.NewSession()
+		results, err := runTests(ctx, logger, "sudo", prefix, script, sshClient, batch.SudoTests)
 		if err != nil {
-			return OSRunnerResult{}, fmt.Errorf("failed to start session: %w", err)
+			return OSRunnerResult{}, fmt.Errorf("error running sudo tests: %w", err)
 		}
+		result.SudoPackages = results
+	}
 
-		session.Stdout = newPrefixOutput(os.Stdout, fmt.Sprintf(">>> (%s) Test output (sudo) (stdout): ", logger.Prefix()))
-		session.Stderr = newPrefixOutput(os.Stderr, fmt.Sprintf(">>> (%s) Test output (sudo) (stderr): ", logger.Prefix()))
-		session.Stdin = execTest
-		// allowed to fail because tests might fail
-		_ = session.Run("bash")
-		_ = session.Close()
+	return result, nil
+}
 
-		// fetch the contents for each package
-		for _, pkg := range batch.SudoTests {
-			resultPkg, err := getRunnerPackageResult(ctx, c, pkg, prefix)
-			if err != nil {
-				return OSRunnerResult{}, err
-			}
-			result.SudoPackages = append(result.SudoPackages, resultPkg)
+func runTests(ctx context.Context, logger Logger, name string, prefix string, script string, sshClient *ssh.Client, tests []define.BatchPackageTests) ([]OSRunnerPackageResult, error) {
+	logger.Logf("starting %s tests", name)
+
+	execTest := strings.NewReader(script)
+
+	logger.Logf("creating new SSH session for %s...", name)
+	session, err := sshClient.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start session: %w", err)
+	}
+
+	session.Stdout = newPrefixOutput(os.Stdout, fmt.Sprintf(">>> (%s) Test output (%s) (stdout): ", name, logger.Prefix()))
+	session.Stderr = newPrefixOutput(os.Stderr, fmt.Sprintf(">>> (%s) Test output (%s) (stderr): ", name, logger.Prefix()))
+	session.Stdin = execTest
+
+	// allowed to fail because tests might fail
+	logger.Logf("running %s tests...", name)
```

Comment:
So, I removed a bunch, but the two in the test runner I'm kind of leaning towards keeping, as in my experience, a lot of the SSH statements tend to sit and hang for an extended period of time, and I figured it would be helpful to have some kind of statement indicating where things are stopping.

Reply:
I am okay with keeping this one.